Sources/PointFree/Admin/AdminMiddleware.swift

```diff
@@ -1,65 +1,66 @@
+import Dependencies
 import Foundation
 import HttpPipeline
 import Models
+import PointFreeDependencies
 import PointFreeRouter
 import Prelude
 import Tuple
 
-let adminMiddleware =
-  requireAdmin
-  <| _adminMiddleware
+public func adminMiddleware(conn: Conn<StatusLineOpen, Admin>) -> IO<Conn<ResponseEnded, Data>> {
+  @Dependency(\.currentUser) var currentUser
+  let route = conn.data
 
-private func _adminMiddleware(
-  conn: Conn<StatusLineOpen, Tuple2<User, Admin>>
-)
-  -> IO<Conn<ResponseEnded, Data>>
-{
+  guard let currentUser = currentUser
+  else { return loginAndRedirect(conn) }
 
-  let (user, route) = lower(conn.data)
+  guard currentUser.isAdmin
+  else {
+    return conn
+      |> redirect(to: .home, headersMiddleware: flash(.error, "You don't have access to that."))
+  }
```

Comment:
Here's an example where I unrolled some middleware transformers and just performed the logic inline. Got rid of `requireAdmin` and instead just unwrap the user and make sure their an admin.

Reply:
I guess `admin: User` could also be a throwing property on `DependencyValues` that throws a different kind of error, which is caught higher up and does the redirect+flash.